File: Technical/3 ClassNotes/Java/src/practiceProblems/OneDeletionEqual.java
package practiceProblems;

public class OneDeletionEqual {

    private static boolean check(String a,
                                 String b,
                                 int i,
                                 int j) {

        while(i < a.length()
                && j < b.length()) {

            if(a.charAt(i)
                    != b.charAt(j))
                return false;

            i++;
            j++;
        }

        return true;
    }

    public static boolean canBecomeEqual(
            String s1,
            String s2) {

        int n1 = s1.length();
        int n2 = s2.length();

        if(Math.abs(n1 - n2) > 1)
            return false;

        int i = 0;
        int j = 0;

        while(i < n1 && j < n2) {

            if(s1.charAt(i)
                    == s2.charAt(j)) {

                i++;
                j++;
            }
            else {

                return check(s1, s2, i + 1, j)
                        || check(s1, s2, i, j + 1);
            }
        }

        return true;
    }

    public static void main(String[] args) {

        System.out.println(
            canBecomeEqual("abcd", "acd")
        );
    }
}

File: Technical/3 ClassNotes/Python/practice_problems/one_deletion_equal.py
def check(a, b, i, j):

    while i < len(a) and j < len(b):

        if a[i] != b[j]:
            return False

        i += 1
        j += 1

    return True


def can_become_equal(s1, s2):

    if abs(len(s1) - len(s2)) > 1:
        return False

    i = j = 0

    while i < len(s1) and j < len(s2):

        if s1[i] == s2[j]:
            i += 1
            j += 1

        else:
            return (
                check(s1, s2, i + 1, j)
                or
                check(s1, s2, i, j + 1)
            )

    return True


print(can_become_equal("abcd", "acd"))